perfkitbenchmarker/providers/ibmcloud/ibmcloud_virtual_machine.py

```diff
@@ -0,0 +1,841 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class to represent a IBM Cloud Virtual Machine. """
+
+import base64
+import json
+import logging
+import threading
+import os
+import time
+import yaml
+import linecache
+
+from subprocess import Popen, PIPE
+
+from absl import flags
+from perfkitbenchmarker import context
+from perfkitbenchmarker import virtual_machine
+from perfkitbenchmarker import linux_virtual_machine
+from perfkitbenchmarker import windows_virtual_machine
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import providers
+from perfkitbenchmarker import vm_util
+from perfkitbenchmarker.providers.ibmcloud import ibmcloud as ibm
+from perfkitbenchmarker.providers.ibmcloud import ibmcloud_disk
+from perfkitbenchmarker.providers.ibmcloud import utils
+
+WINDOWS = 'windows'  # all windows
+UBUNTU = 'ubuntu'  # all ubuntu
+REDHAT = 'redhat'  # all redhat
+DEBIAN = 'debian'  # all debian
+CENTOS = 'centos'  # all centos
+UNKNOWN = 'unknown'
+
+FLAGS = flags.FLAGS
+
+# these are used to create extra subnets for multi vnic support in perfkit,
+# not used for vm provisioning, up to 4 extra totaling 5 subnets
+SUBNET_SUFFIX = 's'
+SUBNET_SUFFIX_EXTRA = 'sxs'
+DELIMITER = 'z-z'
+
+# each prefix range of /18 can only house 4 subnets with /20, use /16, 0: zone1, 1: zone2, etc
+VPC_NAME = 'vpc'
+VPC_PREFIX_RANGES = ['10.101.0.0/16', '10.102.0.0/16', '10.103.0.0/16', '10.104.0.0/16', '10.105.0.0/16']
+
+VPC_SUBNETS = ['10.101.0.0/20', '10.102.0.0/20', '10.103.0.0/20', '10.104.0.0/20', '10.105.0.0/20']
+
+SUBNETX1 = SUBNET_SUFFIX_EXTRA + '1'
+SUBNETX2 = SUBNET_SUFFIX_EXTRA + '2'
+SUBNETX3 = SUBNET_SUFFIX_EXTRA + '3'
+SUBNETX4 = SUBNET_SUFFIX_EXTRA + '4'
+SUBNETXS = [SUBNETX1, SUBNETX2, SUBNETX3, SUBNETX4]
+
+SUBNETS_EXTRA = {
+    SUBNETX1: ['10.101.20.0/24', '10.102.20.0/24', '10.103.20.0/24', '10.104.20.0/24', '10.105.20.0/24'],
+    SUBNETX2: ['10.101.30.0/24', '10.102.30.0/24', '10.103.30.0/24', '10.104.30.0/24', '10.105.30.0/24'],
+    SUBNETX3: ['10.101.40.0/24', '10.102.40.0/24', '10.103.40.0/24', '10.104.40.0/24', '10.105.40.0/24'],
+    SUBNETX4: ['10.101.50.0/24', '10.102.50.0/24', '10.103.50.0/24', '10.104.50.0/24', '10.105.50.0/24']
+}
+
+SUBNETS_EXTRA_GATEWAY = {
+    SUBNETX1: ['10.101.20.1', '10.102.20.1', '10.103.20.1', '10.104.20.1', '10.105.20.1'],
+    SUBNETX2: ['10.101.30.1', '10.102.30.1', '10.103.30.1', '10.104.30.1', '10.105.30.1'],
+    SUBNETX3: ['10.101.40.1', '10.102.40.1', '10.103.40.1', '10.104.40.1', '10.105.40.1'],
+    SUBNETX4: ['10.101.50.1', '10.102.50.1', '10.103.50.1', '10.104.50.1', '10.105.50.1']
+}
+
+USER_DATA = "Content-Type: multipart/mixed; boundary=MIMEBOUNDARY\n\
```

Comment:
What is this USER_DATA used for? Is it necessary?

Reply:
First, thank you for reviewing the changes.
yes, USER_DATA is used for windows virtual machines to set up firewalls and ports used to run benchmarks between vm's.